scrapy_poet/_request_fingerprinter.py

```diff
@@ -0,0 +1,91 @@
+from typing import TYPE_CHECKING
+
+try:
+    from scrapy.utils.request import RequestFingerprinter  # NOQA
+except ImportError:
+    if not TYPE_CHECKING:
+        ScrapyPoetRequestFingerprinter = None
+else:
+    import hashlib
+    import json
+    from functools import cached_property
+    from typing import Callable, Dict, List, Optional
+    from weakref import WeakKeyDictionary
+
+    from scrapy import Request
+    from scrapy.crawler import Crawler
+    from scrapy.settings.default_settings import REQUEST_FINGERPRINTER_CLASS
+    from scrapy.utils.misc import create_instance, load_object
+
+    from scrapy_poet import InjectionMiddleware
+    from scrapy_poet.injection import get_callback
+
+    class ScrapyPoetRequestFingerprinter:
+        @classmethod
+        def from_crawler(cls, crawler):
+            return cls(crawler)
+
+        def __init__(self, crawler: Crawler) -> None:
+            settings = crawler.settings
+            self._base_request_fingerprinter = create_instance(
+                load_object(
+                    settings.get(
+                        "SCRAPY_POET_REQUEST_FINGERPRINTER_BASE_CLASS",
+                        REQUEST_FINGERPRINTER_CLASS,
+                    )
+                ),
+                settings=crawler.settings,
+                crawler=crawler,
+            )
+            self._callback_cache: Dict[Callable, bytes] = {}
+            self._request_cache: "WeakKeyDictionary[Request, bytes]" = (
+                WeakKeyDictionary()
+            )
+            self._crawler: Crawler = crawler
+
+        @cached_property
+        def _injector(self):
+            middlewares = self._crawler.engine.downloader.middleware.middlewares
+            for middleware in middlewares:
+                if isinstance(middleware, InjectionMiddleware):
+                    return middleware.injector
```

Comment:
Unrelated note: I find myself needed to access this as well when debugging. Would it be worth it having this feature in Scrapy itself? Perhaps some getter method in `Crawler` to easily grab some middleware.

Reply:
https://github.com/scrapy/scrapy/pull/6181